src/adaptors/amqp/container.c

```diff
@@ -529,52 +508,43 @@ void qd_container_handle_event(qd_container_t *container, pn_event_t *event,
                 pn_link = pn_link_head(conn, 0);
                 while (pn_link) {
+                    pn_link_t *next_link = pn_link_next(pn_link, 0);
                     if (pn_link_session(pn_link) == ssn) {
-                        qd_link_t *qd_link = (qd_link_t*) pn_link_get_context(pn_link);
-
-                        if ((pn_link_state(pn_link) == (PN_LOCAL_ACTIVE | PN_REMOTE_ACTIVE))) {
-                            if (qd_link) {
-                                if (qd_conn->policy_settings) {
-                                    if (qd_link->direction == QD_OUTGOING) {
-                                        qd_conn->n_receivers--;
-                                        assert(qd_conn->n_receivers >= 0);
-                                    } else {
-                                        qd_conn->n_senders--;
-                                        assert(qd_conn->n_senders >= 0);
-                                    }
+                        qd_link_t *qd_link = (qd_link_t *) pn_link_get_context(pn_link);
+                        if (qd_link) {
+                            if (qd_link->policy_counted) {
+                                qd_link->policy_counted = false;
+                                if (qd_link->direction == QD_OUTGOING) {
+                                    qd_conn->n_receivers--;
+                                    assert(qd_conn->n_receivers >= 0);
+                                } else {
+                                    qd_conn->n_senders--;
+                                    assert(qd_conn->n_senders >= 0);
                                 }
-                                qd_log(LOG_CONTAINER, QD_LOG_DEBUG,
-                                       "Aborting link '%s' due to parent session end", pn_link_name(pn_link));
-                                container->ntype->link_detach_handler(container->qd_router, qd_link, QD_LOST);
                             }
+                            qd_log(LOG_CONTAINER, QD_LOG_DEBUG,
+                                   "Aborting link '%s' due to parent session end", pn_link_name(pn_link));
+                            container->ntype->link_detach_handler(container->qd_router, qd_link, QD_LOST);
+                            qd_link_free(qd_link);
                         }
-
-                        if (qd_link)
-                            qd_link->pn_link = 0;
                     }
-                    pn_link = pn_link_next(pn_link, 0);
-
+                    pn_link = next_link;
                 }
-                if (qd_conn->policy_settings) {
-                    qd_conn->n_sessions--;
-                }
-
                 pn_session_close(ssn);
             }
             else if (pn_session_state(ssn) == (PN_LOCAL_CLOSED | PN_REMOTE_CLOSED)) {
                 qd_session_t *qd_ssn = qd_session_from_pn(ssn);
                 if (qd_ssn) {
+                    if (qd_ssn->policy_counted) {
```

Comment:
Can we please move this `policy_counted` related code to a new commit so this commit only contains the code that fixes the router crash ?

Reply:
I had to add that hacky policy flag because otherwise this patch would break the policy code.  The problem is that not all sessions or links are counted by policy: only those that are 1) remotely initiated and 2) policy is in use.

That means the code cannot _unconditionally_ decrement the relevant policy counters since the counters may not have been incremented for that particular session/link (for reasons 1 and 2 above). So I have to add that ugly "hey I'm counted by policy so remember to decrement when I'm done" hack.

Unconditionally incrementing/decrementing the counters will break the policy logic: there's even asserts in the (existing) code to guard against underflow: https://github.com/skupperproject/skupper-router/blob/2249924bc5304593a71ef36213def662177665cd/src/adaptors/amqp/container.c#L551

Reply:
> That means the code cannot _unconditionally_ decrement the relevant policy counters since the counters may not have been incremented for that particular session/link

Makes sense to me.